Code-4-Community/everything
# feature/monarch/enhanced_distance_filter

### ℹ️ Issue

Closes #2 

### 📝 Description

This change gives visitors of the monarch website the option to filter practitioners by distance. **Importantly, it does not force visitors to filter by distance, which was the previous behavior.** Instead, visitors have the option to enter their zip code. If they enter their zip code, then the website automatically calculates how far away each practitioner is from the visitor and returns the sorted results. If no zip code is entered, then the results will not be filtered by distance.

Changes
- Created a new data definition, GeolocationPosition, in the common monarch directory.
- Added a new GET endpoint /geocode?zipcode=[zipcode] that returns the GeolocationPosition of a zip code provided by the frontend. Modified Zodios definitions and Express workflow to support this change.
- Added a new field, extractGeocodeFromZipcode, to the ActionsController interface definition that outlines a method that the frontend can call to get the geocode from the backend via the controller. Modified actionsController.ts accordingly to support this new method.
- Removed the prompt asking whether the visitor wants to share their location so that there's only 1 way a visitor can share their location (which is through the zip code form field).
- Extracted the language and distance filter into their own components for readability

### ✔️ Verification

- Enable the distance filtering option by toggling the switch.
- Enter a zip code (try zip codes in the New England area such as 02115 and 06110) and verify that the distances calculated by the filtered results seem reasonable.
- Try toggling the distance filter off and attempt to enter another zip code to see whether the behavior is still as expected. When the distance filter is off, or when the user enters an invalid zip code, the results should not be filtered by distance.
- Try to filter by both distance and language. 

![image](https://user-images.githubusercontent.com/73801512/235513148-85a07498-c4ed-43b0-b09d-521de68d3a21.png)

### 🏕️ (Optional) Future Work / Notes

- We should probably perform all filtering operations in the same place. Right now, the distance filtering is being performed in the view (inside of SearchTherapists.tsx) whereas the language filtering is being performed in the controller (inside of actionsController.ts).
- We still need a way to avoid hard-coding the available languages directly into the language filter component.
- Maybe it was not necessary to create a GeolocationPosition definition as there already exists a GeolocationCoordinates data definition from the Geolocation API. The code could be revised to remove this extra definition.
